lib/ut/coroutine2.c

```diff
@@ -0,0 +1,266 @@
+/* -*- C -*- */
+/*
+ * Copyright (c) 2017-2020 Seagate Technology LLC and/or its Affiliates
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+#define M0_TRACE_SUBSYSTEM M0_TRACE_SUBSYS_UT
+#include "lib/trace.h"
+#include "lib/coroutine.h"
+#include "lib/string.h"
+#include "ut/ut.h"
+#include "be/op.h"
+#include "reqh/reqh.h"		/* m0_reqh_fop_handle */
+#include "fop/fom.h"		/* m0_fom_phase_set */
+#include "fop/fom_generic.h"	/* M0_FOPH_FINISH */
+#include "fop/fom_simple.h"
+#include "fop/fom_long_lock.h"
+
+
+extern struct m0_reqh *m0_ut__reqh_init(void);
+extern        void     m0_ut__reqh_fini(void);
+
+enum coroutine2_state {
+	C2_INIT   = M0_FOM_PHASE_INIT,
+	C2_FINISH = M0_FOM_PHASE_FINISH,
+	C2_NR,
+};
+
+static struct m0_sm_state_descr coroutine2_states[C2_NR] = {
+#define _S(name, flags, allowed)      \
+	[name] = {                    \
+		.sd_flags   = flags,  \
+		.sd_name    = #name,  \
+		.sd_allowed = allowed \
+	}
+
+	_S(C2_INIT,   M0_SDF_INITIAL, M0_BITS(C2_INIT, C2_FINISH)),
+	_S(C2_FINISH, M0_SDF_TERMINAL, 0),
+#undef _S
```

Comment:
I am not against this macro defined here.
But, it is only a two-state state descriptor. No need to define such a macro to do the initialization.
We can do explicit initialization, just like @co_states.

Reply:
Initially, it looked the following way:

```+static struct m0_sm_state_descr coroutine2_states[C2_NR] = {
+#define _S(name, flags, allowed)      \
+       [name] = {                    \
+               .sd_flags   = flags,  \
+               .sd_name    = #name,  \
+               .sd_allowed = allowed \
+       }
+
+       _S(C2_INIT,   M0_SDF_INITIAL, M0_BITS(C2_CYCLE, C2_FAILED)),
+       _S(C2_FINISH, M0_SDF_TERMINAL, 0),
+       _S(C2_FAILED, M0_SDF_FAILURE, M0_BITS(C2_FINISH)),
+       _S(C2_CYCLE,  0, M0_BITS(C2_CYCLE, C2_FINISH)),
+#undef _S
+};
```

The growth of code is expected here.
at least 3 states will be added soon:
- cycles,
- jumps to next state (+1)
- jumps to next states (+N)

Reply:
Therefore propose to leave as is it now.

Reply:
OK